gobblin-core/src/main/java/org/apache/gobblin/lineage/LineageInfo.java

```diff
@@ -0,0 +1,218 @@
+   */
+  public static LineageInfo load (State state, Level level) {
+    String datasetUrn = state.getProp(LINEAGE_PREFIX +  ConfigurationKeys.DATASET_URN_KEY, ConfigurationKeys.DEFAULT_DATASET_URN);
+    String jobId = state.getProp(LINEAGE_PREFIX +  ConfigurationKeys.JOB_ID_KEY, "");
+
+    LineageInfo descriptor = new LineageInfo(datasetUrn, jobId);
+
+    for (Map.Entry<Object, Object> entry : state.getProperties().entrySet()) {
+      if (entry.getKey() instanceof String && ((String) entry.getKey()).startsWith(LINEAGE_PREFIX)) {
+        String strippedKey = ((String) entry.getKey()).substring(LINEAGE_PREFIX.length());
+        String value = (String) entry.getValue();
+
+        if (level == Level.TASK_LEVEL) {
+          descriptor.lineageMetaData.put(strippedKey, value);
+        } else if (level == Level.JOB_LEVEL) {
+          if (!strippedKey.startsWith(TASK_PREFIX)) {
+            descriptor.lineageMetaData.put(strippedKey, value);
+          }
+        } else if (level == Level.TASK_LEVEL_ONLY) {
+          if (strippedKey.startsWith(TASK_PREFIX)) {
+            descriptor.lineageMetaData.put(strippedKey, value);
+          }
+        }
+      }
+    }
+
+    return descriptor;
+  }
+
+  /**
+   * Get all lineage attributes. All LINEAGE_PREFIX will be stripped.
+   */
+  public ImmutableMap<String, String> getLineageMetaData() {
+    return ImmutableMap.copyOf(lineageMetaData);
+  }
+
+  /**
+   * Load shared lineage attributes.
+   */
+  public static LineageInfo load (Collection<? extends State> states, Level level ) {
+    Preconditions.checkState(!states.isEmpty());
+    State s = states.iterator().next();
+
+    // datasetUrn and jobId is a job level attribute
+    String datasetUrn = s.getProp(LINEAGE_PREFIX +  ConfigurationKeys.DATASET_URN_KEY, ConfigurationKeys.DEFAULT_DATASET_URN);
+    String jobId = s.getProp(LINEAGE_PREFIX +  ConfigurationKeys.JOB_ID_KEY, "");
+
+    LineageInfo descriptor = new LineageInfo(datasetUrn, jobId);
+    Map<String, Set<String>> jobLevel = new HashMap<>();
+
+    for (State state: states) {
+      for (Map.Entry<Object, Object> entry : state.getProperties().entrySet()) {
+        if (entry.getKey() instanceof String && ((String) entry.getKey()).startsWith(LINEAGE_PREFIX)) {
+          String strippedKey = ((String) entry.getKey()).substring(LINEAGE_PREFIX.length());
+          String value = (String) entry.getValue();
+
+          if (level == Level.JOB_LEVEL) {
+            if (!strippedKey.startsWith(TASK_PREFIX)) {
+              if (!jobLevel.containsKey(strippedKey)) {
+                jobLevel.put(strippedKey, new HashSet<>());
+              }
+              jobLevel.get(strippedKey).add(value);
+            }
+          } else {
+            throw new UnsupportedOperationException("only job level aggregation is supported, will consider task level support int he future.");
+          }
+        }
+      }
+    }
+
+    for (Map.Entry<String, Set<String>> entry: jobLevel.entrySet()) {
```

Comment:
If you are going to use only one, there is no point using a set above. You can just compare against the current value, and if it is different, throw an exception early at that point.